cmd/egress-v6-cni-plugin/cni.go

```diff
@@ -0,0 +1,608 @@
+			}
+		}
+		return os.WriteFile(entry, []byte("2"), 0644)
+	} else {
+		return fmt.Errorf("failed to get host primary interface name with mac: %s", primaryMAC)
+	}
+}
+
+func setupHostIPv6Route(netns ns.NetNS, containerVethIfName string, containerIPv6 net.IP) error {
+
+	hostVethIfIndex, err := getHostVethIfIndex(netns, containerVethIfName)
+	if err != nil {
+		return err
+	}
+	// set up to container traffic route
+	return netlink.RouteAdd(&netlink.Route{
+		LinkIndex: *hostVethIfIndex,
+		Scope:     netlink.SCOPE_HOST,
+		Dst:       &net.IPNet {
+			IP: containerIPv6,
+			Mask: net.CIDRMask(128, 128),
+		},
+	})
+}
+
+func tearDownHostIPv6Route(netns ns.NetNS, containerVethIfName string, containerIPv6 net.IP) error {
+	hostVethIfIndex, err := getHostVethIfIndex(netns, containerVethIfName)
+	if err != nil {
+		return err
+	}
+	// set up to container traffic route
+	return netlink.RouteDel(&netlink.Route{
+		LinkIndex: *hostVethIfIndex,
+		Scope:     netlink.SCOPE_HOST,
+		Dst:       &net.IPNet {
+			IP: containerIPv6,
+			Mask: net.CIDRMask(128, 128),
+		},
+	})
+}
+
+func setupHostIPv6Snat(containerIPv6 net.IP, hostPrimaryIfIPv6 net.IP, chain, comment, randomizeSNAT string) error {
+	return  snat.Snat6(hostPrimaryIfIPv6, containerIPv6, chain, comment, randomizeSNAT)
+}
+
+func main() {
+	skel.PluginMain(cmdAdd, nil, cmdDel, cniversion.All, fmt.Sprintf("egress-v6 CNI plugin %s", version))
+}
+
+func cmdAdd(args *skel.CmdArgs) error {
+	netConf, log, err := loadConf(args.StdinData)
+	if err != nil {
+		log.Debugf("Received Add request: Failed to parse config")
+		return fmt.Errorf("failed to parse config: %v", err)
+	}
+
+	if netConf.PrevResult == nil {
+		return fmt.Errorf("must be called as a chained plugin")
+	}
+
+	result, err := current.GetResult(netConf.PrevResult)
+	if err != nil {
+		return err
+	}
+
+	log.Debugf("Received an ADD request for: conf=%v; Plugin enabled=%s", netConf, netConf.Enabled)
+	log.Debugf("PrevResult: %v", result)
+	// We will not be vending out this as a separate plugin by itself, and it is only intended to be used as a
+	// chained plugin to VPC CNI in IPv4 mode.
+	// We only need this plugin to kick in if v6 is NOT enabled in VPC CNI. So, the
+	// value of an env variable in VPC CNI determines whether this plugin should be enabled and this is an attempt to
+	// pass through the variable configured in VPC CNI.
+	if netConf.Enabled == "false" {
+		return types.PrintResult(result, netConf.CNIVersion)
+	}
+
+	hostPrimaryIfIPv6AddrStr, err := imds.GetMetaData("ipv6")
+	if err != nil {
+		log.Debugf("IPv6 address not found on host primary interface which is needed for IPv6 egress: %v", err)
+		return err
+	}
+
+	hostPrimaryIfIPv6 := net.ParseIP(hostPrimaryIfIPv6AddrStr)
+	if hostPrimaryIfIPv6 == nil || !hostPrimaryIfIPv6.IsGlobalUnicast() {
+		return fmt.Errorf("gobal unicast IPv6 address is needed on host primary interface for IPv6 egress")
+	}
+	log.Debugf("IPv6 address retrieved: %s", hostPrimaryIfIPv6AddrStr)
```

Comment:
The IPv6 address assigned to the primary ENI should be derived during container startup and then set in `NetConf.NodeIP`: https://github.com/aws/amazon-vpc-cni-k8s/blob/master/cmd/aws-vpc-cni/main.go#L389 , similar to what is done for IPv4 egress.

The reason I think we should do this during container startup is that if the customer does not configure an IPv6 address on the primary ENI, we can log an error early and set the chained plugin to `Disabled`. That way, the plugin will never be enabled unless the customer config meets that precondition.

For resiliency, if the IPv6 address happens to be removed on the primary ENI while the plugin is running, the result should be that we just create an SNAT rule that won't actually forward. And have no crash.